mirrord/cli/src/vpn.rs

```diff
@@ -0,0 +1,302 @@
+        .inspect_err(|_| analytics.set_error(AnalyticsError::AgentConnection))?;
+
+    sub_progress.success(None);
+
+    let Some(ServerVpn::NetworkConfiguration(network)) = agent_request(
+        &mut connection,
+        ClientMessage::Vpn(ClientVpn::GetNetworkConfiguration),
+        get_server_vpn,
+    )
+    .await
+    else {
+        return Ok(());
+    };
+
+    tracing::debug!(?network, "loaded vpn network configuration");
+
+    let mut sub_progress = progress.subtask("create tun socket");
+
+    let (mut write_stream, read_stream) = mirrord_vpn::socket::create_vpn_socket(&network).split();
+    let read_stream = read_stream.fuse();
+    tokio::pin!(read_stream);
+
+    sub_progress.success(None);
+
+    #[cfg(not(target_os = "macos"))]
+    let linux_guard = {
+        use mirrord_vpn::linux::*;
+
+        let Ok(remote_resolv) =
+            agent_fetch_file::<10000>(&mut connection, "/etc/resolv.conf".into()).await
+        else {
+            return Ok(());
+        };
+
+        let Ok(resolv_override) = ResolvOverride::accuire_override("/etc/resolv.conf")
+            .await
+            .inspect_err(|error| tracing::info!(%error, "unabable to override /etc/resolv.conf"))
+        else {
+            return Ok(());
+        };
+
+        if resolv_override.update_resolv(&remote_resolv).await.is_err() {
+            let _ = resolv_override.unmount().await;
+
+            return Ok(());
+        }
+
+        let _ = tokio::process::Command::new("ip")
+            .args([
+                "route".to_owned(),
+                "add".to_owned(),
+                vpn_config.service_subnet.to_string(),
+                "via".to_owned(),
+                network.gateway.to_string(),
+            ])
+            .output()
+            .await
+            .inspect_err(|error| tracing::error!(%error, "could not bind service_subnet"));
+
+        resolv_override
+    };
+
+    #[cfg(target_os = "macos")]
+    let macos_guard = {
+        use mirrord_vpn::macos::*;
+
+        let subnet_guard = create_subnet_route(&vpn_config.service_subnet, &network.gateway)
+            .await
+            .map_err(CliError::RuntimeError)?;
+
+        let resolve_guard = ResolveFile {
+            port: 53,
+            domain: vpn_config.dns_domain,
+            nameservers: vpn_config.dns_nameservers,
+            ..Default::default()
+        }
+        .inject()
+        .await
+        .map_err(CliError::RuntimeError)?;
+
+        (subnet_guard, resolve_guard)
+    };
+
+    let packets_sent = AtomicUsize::default();
+    let bytes_sent = AtomicUsize::default();
+
+    let packets_recived = AtomicUsize::default();
+    let bytes_recived = AtomicUsize::default();
+
+    let mut statistic_interval = tokio::time::interval(Duration::from_secs(10));
+
+    progress.success(None);
+
+    let _ = connection
+        .sender
+        .send(ClientMessage::Vpn(ClientVpn::OpenSocket))
+        .await;
+
+    'main: loop {
+        tokio::select! {
```

Comment:
There are some `unwrap`s and `unimplemented` in this loop. If they happen, cleanups (starting at line 288) won't run